frontend_server/lib/server.dart

```diff
@@ -309,9 +291,9 @@ Future<int> starter(
           boundaryKey = string.substring(RECOMPILE_INSTRUCTION_SPACE.length);
           state = _State.RECOMPILE_LIST;
         } else if (string == 'accept') {
-          compiler.acceptLastDelta();
+          // TODO(vegorov) remove this command from the protocol
         } else if (string == 'reject') {
-          compiler.rejectLastDelta();
+          // TODO(vegorov) remove this command from the protocol
```

Comment:
Curious, how does incremental compiler work without `rejectLastDelta`?
How does incremental compiler know when to included rejected changes(i.e. changes that VM could not load) into next incremental kernel file?

Reply:
I guess now it this does not necessarily work. Are there any tests or design documentation for this behavior? I would like to learn what sort of changes can be rejected by the VM and how unrolling front-end state fixes it.

/cc @peter-ahe-google

Reply:
For now, please create a new instance. I don't understand why the incremental compiler needs to keep track of what VM doesn't like, but I'll be happy to help you move this kind of logic to package/vm.

Reply:
Implemented reset on reject.

Reply:
The best documentation I can think of is this: https://github.com/flutter/engine/tree/master/frontend_server#acceptreject

I'm thinking about class->enum kind of changes that current VM hot reload won't accept. Not sure what else VM hot reload would not be happy with. @rmacnak-google can comment.

In previous IKG concept was that frontend needs to include files that resulted in rejected hot reload request so that VM sees complete incremental picture since last time incremental file was 
accepted.

Reply:
Peter, I don't think the incremental compiler needs to know what kind of changes the VM doesn't support. It does need to know what is currently loaded into the VM to produce a correct delta, and needs to be aware that the most recent delta it produced might not have been loaded into the VM (due to an IO error or a kind of change the VM doesn't know how to handle), but it need not try to eagerly detect the extra restrictions the VM adds.

Slava, see callers of AddReasonForCancelling for the kinds of changes the VM does not support. This delta tracking business should allow someone to make a kind of change the VM does not support, have the change be rejected, then make a different kind of change the VM does support and have the second change be properly loaded as if the first was never attempted.

Reply:
@rmacnak-google makes total sense. Thanks for clarifications. 

I think it would be great to ensure that we have coverage for these features somewhere in the test suite (does not matter if in Flutter/Flutter Engine or SDK repos). 

We are gonna implement rejection handling in the separate CL on top of public incremental compilation APIs.

Reply:
I'll be happy to help getting rejection working on top of the front-end API.